src/Compilers/Core/Portable/Compilation/CompilationOptions.cs

```diff
@@ -466,6 +471,47 @@ public CompilationOptions WithStrongNameProvider(StrongNameProvider provider)
         /// </summary>
         internal abstract void ValidateOptions(ArrayBuilder<Diagnostic> builder);
 
+        internal void ValidateOptions(ArrayBuilder<Diagnostic> builder, CommonMessageProvider messageProvider)
+        {
+            if (!CryptoPublicKey.IsEmpty)
+            {
+                if (CryptoKeyFile != null)
+                {
+                    builder.Add(messageProvider.CreateDiagnostic(messageProvider.ERR_MutuallyExclusiveOptions,
+                        Location.None, nameof(CryptoPublicKey), nameof(CryptoKeyFile)));
+                }
+
+                if (CryptoKeyContainer != null)
+                {
+                    builder.Add(messageProvider.CreateDiagnostic(messageProvider.ERR_MutuallyExclusiveOptions,
+                        Location.None, nameof(CryptoPublicKey), nameof(CryptoKeyContainer)));
+                }
+            }
+
+            if (PublicSign)
+            {
+                if (CryptoKeyFile != null && !PathUtilities.IsAbsolute(CryptoKeyFile))
+                {
+                    // TODO(https://github.com/dotnet/roslyn/issues/9153):
+                    // Produce better diagnostic message for passing a key file with a relative path
+                    builder.Add(messageProvider.CreateDiagnostic(messageProvider.ERR_BadCompilationOptionValue,
+                        Location.None, nameof(CryptoKeyFile), CryptoKeyFile));
+                }
+
+                if (CryptoKeyContainer != null)
+                {
+                    builder.Add(messageProvider.CreateDiagnostic(messageProvider.ERR_MutuallyExclusiveOptions,
```

Comment:
C# already had the test, called `KeyContainerNoSNProvider_PublicSign`. It looks like VB did not have one, so I added one called `PublicSign_KeyContainerOnly`.